api/app/controllers/spree/api/base_controller.rb

```diff
@@ -2,170 +2,176 @@
+        end.with_indifferent_access
+      end
+
+      # users should be able to set price when importing orders via api
+      def permitted_line_item_attributes
+        if @current_user_roles.include?("admin")
+          super + [:price, :variant_id, :sku]
+        else
+          super
         end
+      end
 
-        def unauthorized
-          render "spree/api/errors/unauthorized", status: 401 and return
+      private
+
+      def content_type
+        case params[:format]
+        when "json"
+          "application/json; charset=utf-8"
+        when "xml"
+          "text/xml; charset=utf-8"
         end
+      end
 
-        def error_during_processing(exception)
-          Rails.logger.error exception.message
-          Rails.logger.error exception.backtrace.join("\n")
+      def set_content_type
+        headers["Content-Type"] = content_type
+      end
 
-          error_notifier.call(exception, self) if error_notifier
+      def load_user
+        @current_api_user = (try_spree_current_user || Spree.user_class.find_by(spree_api_key: api_key.to_s))
+      end
 
-          render text: { exception: exception.message }.to_json,
-            status: 422 and return
-        end
+      def authenticate_user
+        return if @current_api_user
 
-        def gateway_error(exception)
-          @order.errors.add(:base, exception.message)
-          invalid_resource!(@order)
+        if requires_authentication? && api_key.blank? && order_token.blank?
+          render "spree/api/errors/must_specify_api_key", status: 401 and return
+        elsif order_token.blank? && (requires_authentication? || api_key.present?)
+          render "spree/api/errors/invalid_api_key", status: 401 and return
+        else
+          # An anonymous user
+          @current_api_user = Spree.user_class.new
         end
+      end
 
-        def requires_authentication?
-          Spree::Api::Config[:requires_authentication]
-        end
+      def load_user_roles
+        @current_user_roles = @current_api_user ? @current_api_user.spree_roles.pluck(:name) : []
+      end
 
-        def not_found
-          render "spree/api/errors/not_found", status: 404 and return
-        end
+      def unauthorized
+        render "spree/api/errors/unauthorized", status: 401 and return
+      end
 
-        def current_ability
-          Spree::Ability.new(current_api_user)
-        end
+      def error_during_processing(exception)
+        Rails.logger.error exception.message
+        Rails.logger.error exception.backtrace.join("\n")
 
-        def current_currency
-          Spree::Config[:currency]
-        end
-        helper_method :current_currency
+        error_notifier.call(exception, self) if error_notifier
 
-        def invalid_resource!(resource)
-          @resource = resource
-          render "spree/api/errors/invalid_resource", :status => 422
-        end
+        render text: { exception: exception.message }.to_json,
+          status: 422 and return
```

Comment:
Align the elements of a hash literal if they span more than one line.<br>Use `&&` instead of `and`.